src/sagemaker_xgboost_container/metrics/custom_metrics.py

```diff
@@ -23,7 +36,7 @@ def accuracy(preds, dtrain):
     :return: Metric name, accuracy value.
     """
     labels = dtrain.get_label()
-    rounded_preds = [np.argmax(value) if (type(value) is np.ndarray) else round(value) for value in preds]
+    rounded_preds = [np.argmax(x) if (type(x) is np.ndarray) else round(sigmoid(x+0.5)) for x in preds]
```

Comment:
Would it be reasonable to just replace the `round(value)` with `int(value > 0.5)`? 

That way it works with both probabilities and with log-odds. Additionally we don't have to do the extra transformation (which could end up being numerically unstable).

Reply:
log-odds doesn't lie in [0,1].

Reply:
yes, logodds is [-infinity, infinity] but it doesn't matter here since we're mapping that to `0` and `1` for f1 score. Basically, instead of doing the comparison in probability space (i.e. check if `prob > 0.5`) we compare in logodds space (i.e. check if `logodds > 0`). The `+0.5` that you're doing is for incorporating the base_margin which is 0.5 by default i.e treat random classifier as the base. To take that into account we have to compare logodds with `-0.5` [above comment comparing with `0.5` is wrong - it should say `int(value > -0.5)`]. 

I've the numbers below for binary classification in Adult Census data using the various objectives. 

# binary:logistic

## round(value) [this one fails to give correct f1 score]
```
[0]	dtrain-error:0.16013	dtrain-f1:0.42131
[1]	dtrain-error:0.16053	dtrain-f1:0.42131
[2]	dtrain-error:0.16093	dtrain-f1:0.42131
[3]	dtrain-error:0.16117	dtrain-f1:0.14953
[4]	dtrain-error:0.16112	dtrain-f1:0.14351
[5]	dtrain-error:0.16152	dtrain-f1:0.21430
[6]	dtrain-error:0.16043	dtrain-f1:0.26171
[7]	dtrain-error:0.16043	dtrain-f1:0.28508
[8]	dtrain-error:0.16043	dtrain-f1:0.31020
[9]	dtrain-error:0.16043	dtrain-f1:0.23860
[10]	dtrain-error:0.16043	dtrain-f1:0.23927
[11]	dtrain-error:0.15781	dtrain-f1:0.24134
[12]	dtrain-error:0.15905	dtrain-f1:0.17392
[13]	dtrain-error:0.15905	dtrain-f1:0.17420
[14]	dtrain-error:0.15825	dtrain-f1:0.17080
[15]	dtrain-error:0.15830	dtrain-f1:0.17055
[16]	dtrain-error:0.15603	dtrain-f1:0.17122
[17]	dtrain-error:0.15395	dtrain-f1:0.17154
[18]	dtrain-error:0.15425	dtrain-f1:0.17131
[19]	dtrain-error:0.15420	dtrain-f1:0.18879
[20]	dtrain-error:0.15415	dtrain-f1:0.18923
[21]	dtrain-error:0.15405	dtrain-f1:0.18887
[22]	dtrain-error:0.15415	dtrain-f1:0.13167
[23]	dtrain-error:0.15429	dtrain-f1:0.13187
[24]	dtrain-error:0.15439	dtrain-f1:0.12226
```

## int(value > -0.5) [this one gives increasing f1 score]
```
[0]	dtrain-error:0.16013	dtrain-f1:0.21382
[1]	dtrain-error:0.16053	dtrain-f1:0.21382
[2]	dtrain-error:0.16093	dtrain-f1:0.21382
[3]	dtrain-error:0.16117	dtrain-f1:0.68871
[4]	dtrain-error:0.16112	dtrain-f1:0.69781
[5]	dtrain-error:0.16152	dtrain-f1:0.69781
[6]	dtrain-error:0.16043	dtrain-f1:0.69749
[7]	dtrain-error:0.16043	dtrain-f1:0.69656
[8]	dtrain-error:0.16043	dtrain-f1:0.69656
[9]	dtrain-error:0.16043	dtrain-f1:0.77968
[10]	dtrain-error:0.16043	dtrain-f1:0.77968
[11]	dtrain-error:0.15781	dtrain-f1:0.77968
[12]	dtrain-error:0.15905	dtrain-f1:0.78977
[13]	dtrain-error:0.15905	dtrain-f1:0.78977
[14]	dtrain-error:0.15825	dtrain-f1:0.79226
[15]	dtrain-error:0.15830	dtrain-f1:0.79245
[16]	dtrain-error:0.15603	dtrain-f1:0.79239
[17]	dtrain-error:0.15395	dtrain-f1:0.79263
[18]	dtrain-error:0.15425	dtrain-f1:0.79295
[19]	dtrain-error:0.15420	dtrain-f1:0.79277
[20]	dtrain-error:0.15415	dtrain-f1:0.79334
[21]	dtrain-error:0.15405	dtrain-f1:0.79356
[22]	dtrain-error:0.15415	dtrain-f1:0.79321
[23]	dtrain-error:0.15429	dtrain-f1:0.79466
[24]	dtrain-error:0.15439	dtrain-f1:0.79471
```

## round(sigmoid(value + 0.5)) [these values are same as the one above]
``` 
[0]	dtrain-error:0.16013	dtrain-f1:0.21382
[1]	dtrain-error:0.16053	dtrain-f1:0.21382
[2]	dtrain-error:0.16093	dtrain-f1:0.21382
[3]	dtrain-error:0.16117	dtrain-f1:0.68871
[4]	dtrain-error:0.16112	dtrain-f1:0.69781
[5]	dtrain-error:0.16152	dtrain-f1:0.69781
[6]	dtrain-error:0.16043	dtrain-f1:0.69749
[7]	dtrain-error:0.16043	dtrain-f1:0.69656
[8]	dtrain-error:0.16043	dtrain-f1:0.69656
[9]	dtrain-error:0.16043	dtrain-f1:0.77968
[10]	dtrain-error:0.16043	dtrain-f1:0.77968
[11]	dtrain-error:0.15781	dtrain-f1:0.77968
[12]	dtrain-error:0.15905	dtrain-f1:0.78977
[13]	dtrain-error:0.15905	dtrain-f1:0.78977
[14]	dtrain-error:0.15825	dtrain-f1:0.79226
[15]	dtrain-error:0.15830	dtrain-f1:0.79245
[16]	dtrain-error:0.15603	dtrain-f1:0.79239
[17]	dtrain-error:0.15395	dtrain-f1:0.79263
[18]	dtrain-error:0.15425	dtrain-f1:0.79295
[19]	dtrain-error:0.15420	dtrain-f1:0.79277
[20]	dtrain-error:0.15415	dtrain-f1:0.79334
[21]	dtrain-error:0.15405	dtrain-f1:0.79356
[22]	dtrain-error:0.15415	dtrain-f1:0.79321
[23]	dtrain-error:0.15429	dtrain-f1:0.79466
[24]	dtrain-error:0.15439	dtrain-f1:0.79471
```

# binary:logitraw

## int(value > -0.5)  [this works for logitraw]
```
[0]	dtrain-auc:0.84994	dtrain-f1:0.21382
[1]	dtrain-auc:0.85610	dtrain-f1:0.21382
[2]	dtrain-auc:0.86501	dtrain-f1:0.21382
[3]	dtrain-auc:0.86624	dtrain-f1:0.68871
[4]	dtrain-auc:0.87011	dtrain-f1:0.69781
[5]	dtrain-auc:0.87526	dtrain-f1:0.69781
[6]	dtrain-auc:0.87557	dtrain-f1:0.69749
[7]	dtrain-auc:0.88166	dtrain-f1:0.69656
[8]	dtrain-auc:0.88177	dtrain-f1:0.69656
[9]	dtrain-auc:0.88157	dtrain-f1:0.77968
[10]	dtrain-auc:0.88190	dtrain-f1:0.77968
[11]	dtrain-auc:0.88875	dtrain-f1:0.77968
[12]	dtrain-auc:0.88924	dtrain-f1:0.78977
[13]	dtrain-auc:0.89125	dtrain-f1:0.78977
[14]	dtrain-auc:0.89779	dtrain-f1:0.79226
[15]	dtrain-auc:0.89898	dtrain-f1:0.79245
[16]	dtrain-auc:0.89914	dtrain-f1:0.79239
[17]	dtrain-auc:0.89988	dtrain-f1:0.79263
[18]	dtrain-auc:0.90169	dtrain-f1:0.79295
[19]	dtrain-auc:0.90315	dtrain-f1:0.79277
[20]	dtrain-auc:0.90391	dtrain-f1:0.79334
[21]	dtrain-auc:0.90503	dtrain-f1:0.79356
[22]	dtrain-auc:0.90483	dtrain-f1:0.79321
[23]	dtrain-auc:0.90575	dtrain-f1:0.79466
[24]	dtrain-auc:0.90542	dtrain-f1:0.79471
```

# binary:hinge

## int(value > -0.5) [... and it works for hinge]
```
[0]	dtrain-error:0.72803	dtrain-f1:0.21382
[1]	dtrain-error:0.72803	dtrain-f1:0.21382
[2]	dtrain-error:0.72803	dtrain-f1:0.21382
[3]	dtrain-error:0.72803	dtrain-f1:0.21382
[4]	dtrain-error:0.72803	dtrain-f1:0.21382
[5]	dtrain-error:0.28677	dtrain-f1:0.21382
[6]	dtrain-error:0.28677	dtrain-f1:0.21382
[7]	dtrain-error:0.28677	dtrain-f1:0.21382
[8]	dtrain-error:0.28558	dtrain-f1:0.21382
[9]	dtrain-error:0.28558	dtrain-f1:0.21382
[10]	dtrain-error:0.28558	dtrain-f1:0.21382
[11]	dtrain-error:0.28553	dtrain-f1:0.69025
[12]	dtrain-error:0.15880	dtrain-f1:0.69025
[13]	dtrain-error:0.15410	dtrain-f1:0.69956
[14]	dtrain-error:0.15410	dtrain-f1:0.69988
[15]	dtrain-error:0.15410	dtrain-f1:0.69988
[16]	dtrain-error:0.15415	dtrain-f1:0.69988
[17]	dtrain-error:0.15375	dtrain-f1:0.69988
[18]	dtrain-error:0.15375	dtrain-f1:0.70028
[19]	dtrain-error:0.15360	dtrain-f1:0.70125
[20]	dtrain-error:0.15415	dtrain-f1:0.70119
[21]	dtrain-error:0.15375	dtrain-f1:0.70140
[22]	dtrain-error:0.15375	dtrain-f1:0.70144
[23]	dtrain-error:0.15345	dtrain-f1:0.78502
[24]	dtrain-error:0.15291	dtrain-f1:0.78502
```

Reply:
I tried but something is wrong. I include accuracy in the eval_metric and find that `accuracy + error != 1`. Not sure what I missed. Can you try on that too?

BTW, why do you check `preds.size > 0`? In what case, xgb will pass empty prediction?